### DISCUSSION TITLE
different behaviour with types instanciation in last update

### DISCUSSION THREAD

**Discussion by crissi** [rebing/graphql-laravel#789] [Q&A] [answered]
### Versions:
- graphql-laravel Version: 8.0.0-rc3
- Laravel Version: 8.0.0
- PHP Version: 7.4

### Description:
I somehow did not catch this in my testing. Not sure this is webonyx or the latest changes here.

Previous releases, types were instanciated each time a new query were sent.

### Steps To Reproduce:
I am instanciating a buffer in a type.

![image](https://user-images.githubusercontent.com/3076578/120299431-482d2580-c2cb-11eb-8ce0-beb0d69cc9de.png)

And using that. If I am batching two queries that use the same type or just running tests where am quering the same query multiple times in a row. 

This Buffer will only be instanciated once.

Am I doing the instanciating wrong, are you doing it differently?

**Comment by crissi** [comment]
should have labeled as a question

**Reply by mfn** [reply]
> should have labeled as a question

I think I should remove this label and encourage discussion (I just converted the issue 😄 ).

**Comment by mfn** [comment, marked as answer]
I yet have to analyze the current behaviour or change of behaviour etc.

But from my POV, nothing non-GraphQL related should be instantiated in `fields` or `args`, etc. My argument would be: these methods are a fancy way to describe a DSL and usually you don't run application instantiation related code there.

If you need access to this stuff in a resolver, the way to go would be to use the context which is essentially "a bag of things" which are passed around and you mutate as you want (dangerously like a "global").

---
That said: I'm now curious what your actual use case here is 😄

**Reply by mfn** [reply]
> and usually you don't run application instantiation related code there.

To expand this: ideally we would strive to build the GraphQL schema only once and never again (think laravel/octane).

**Reply by crissi** [reply]
I have an graphql endpoint that returns multiple entries and I want to fire one sql query for all them. So I collect all entries and fire one sql afterwards using this logic: https://github.com/webonyx/graphql-php/blob/master/docs/data-fetching.md

But I have always initiated my buffer in fields.

So how do you use context?

Override this GraphQLController::queryContext and set it to an instance of a class!?

And then use a set method on it in the query resolver?

**Reply by crissi** [reply]
Ahh I see it is now a middleware:)

**Reply by mfn** [reply]
Yes, it's now a middleware. A better design than we had before, unfortunately a breaking change.

But yes, the data loader concept is the way to go and the "context" is the place to share it along all your resolvers 👍

**Comment by crissi** [comment]
I would have to initialize every buffer in middleware or try to dig into the query response and see if there is type that needs a buffer or how are you doing it?

**Reply by crissi** [reply]
okay I think I got it. Would generate a generic buffer that I could use where I want to

**Reply by mfn** [reply]
They way I use those buffers / data loaders it that:
- I've a context which "can carry arbitrary buffers"
- when I've a resolver which needs such a buffer, I register it with the context
- when it's requested again in another resolver, it's "shared" (aka not register new again but retrieved) due to the same context

Essentially the context acts like global storage for the lifecycle of this GraphQL execution.

**Reply by crissi** [reply]
I see, I like it 👍
